File: server/game/GameManager.js
/**
 * Manages the overall game flow, rules, and game modes.
 */
class GameManager {
  /**
   * Creates a new game manager.
   */
  constructor() {
    // Game settings
    this.gameTimeLimit = 600; // 10 minutes in seconds
    this.maxPlayers = 8;
    
    // Game state
    this.gameTimer = 0;
    this.playersEscaped = 0;
    this.playersTrapped = 0;
    this.isGameOver = false;
    this.gameMode = 'coop'; // coop, betrayal, pvp
    
    // Player management
    this.players = new Map();
    
    // Maze reference
    this.maze = null;
  }
  
  /**
   * Updates the game timer.
   * @param {number} deltaTime - Time since last update in milliseconds
   */
  updateGameTimer(deltaTime) {
    if (this.isGameOver) return;
    
    this.gameTimer += deltaTime / 1000; // Convert to seconds
    
    // Check for time-based game over
    if (this.gameTimer >= this.gameTimeLimit) {
      this.checkGameOverConditions();
    }
  }
  
  /**
   * Adds a new player to the game.
   * @param {string} playerId - The player's socket ID
   * @returns {Object} The created player object
   */
  addPlayer(playerId) {
    // Create a new player object
    const player = {
      id: playerId,
      position: this.getRandomSpawnPosition(),
      health: 100,
      hasEscaped: false,
      isTrapped: false,
      inventory: []
    };
    
    // Add to players map
    this.players.set(playerId, player);
    
    return player;
  }
  
  /**
   * Removes a player from the game.
   * @param {string} playerId - The player's socket ID
   */
  removePlayer(playerId) {
    this.players.delete(playerId);
  }
  
  /**
   * Gets a random spawn position for a new player.
   * @returns {Object} A position object with x, y, z coordinates
   */
  getRandomSpawnPosition() {
    // If no maze yet, return a default position
    if (!this.maze) {
      return { x: 0, y: 0, z: 0 };
    }
    
    // Find a random path cell that's not the exit
    const validCells = [];
    for (let x = 0; x < this.maze.width; x++) {
      for (let y = 0; y < this.maze.height; y++) {
        if (!this.maze.grid[x][y] && 
            (x !== this.maze.exitPosition.x || y !== this.maze.exitPosition.y)) {
          validCells.push({ x, y });
        }
      }
    }
    
    if (validCells.length === 0) {
      return { x: 0, y: 0, z: 0 };
    }
    
    // Pick a random cell
    const cell = validCells[Math.floor(Math.random() * validCells.length)];
    
    // Convert to world position
    return {
      x: cell.x * this.maze.cellSize,
      y: 0, // Ground level
      z: cell.y * this.maze.cellSize
    };
  }
  
  /**
   * Updates a player's position.
   * @param {string} playerId - The player's socket ID
   * @param {Object} position - The new position
   */
  updatePlayerPosition(playerId, position) {
    const player = this.players.get(playerId);
    if (player) {
      player.position = position;
    }
  }
  
  /**
   * Handles a player's interaction with the environment.
   * @param {string} playerId - The player's socket ID
   * @returns {Object} The result of the interaction
   */
  handlePlayerInteraction(playerId) {
    const player = this.players.get(playerId);
    if (!player) {
      return { success: false, message: 'Player not found' };
    }
    
    // Check if player is near the exit
    if (this.isPlayerNearExit(player)) {
      return {
        success: true,
        type: 'exit',
        message: 'Press E to escape'
      };
    }
    
    // Check for items nearby
    const item = this.findItemNearPlayer(player);
    if (item) {
      // Add item to player's inventory
      player.inventory.push(item);
      
      return {
        success: true,
        type: 'item',
        item: item,
        message: `Picked up ${item.name}`
      };
    }
    
    return { success: false, message: 'Nothing to interact with' };
  }
  
  /**
   * Checks if a player is near the exit.
   * @param {Object} player - The player object
   * @returns {boolean} True if the player is near the exit
   */
  isPlayerNearExit(player) {
    if (!this.maze || !this.maze.exitPosition) return false;
    
    // Convert exit position to world coordinates
    const exitWorldPos = {
      x: this.maze.exitPosition.x * this.maze.cellSize,
      z: this.maze.exitPosition.y * this.maze.cellSize
    };
    
    // Calculate distance
    const distance = Math.sqrt(
      Math.pow(player.position.x - exitWorldPos.x, 2) +
      Math.pow(player.position.z - exitWorldPos.z, 2)
    );
    
    return distance < 3; // Within 3 units of the exit
  }
  
  /**
   * Finds an item near a player.
   * @param {Object} player - The player object
   * @returns {Object|null} The item if found, null otherwise
   */
  findItemNearPlayer(player) {
    // This would be implemented with actual item spawning logic
    // For now, just return null (no items)
    return null;
  }
  
  /**
   * Handles a player escaping the maze.
   * @param {string} playerId - The player's socket ID
   * @returns {Object} The result of the escape attempt
   */
  playerEscape(playerId) {
    const player = this.players.get(playerId);
    if (!player) {
      return { escaped: false, message: 'Player not found' };
    }
    
    // Check if player is already escaped
    if (player.hasEscaped) {
      return { escaped: false, message: 'Player already escaped' };
    }
    
    // Check if player is near the exit
    if (!this.isPlayerNearExit(player)) {
      return { escaped: false, message: 'Not near the exit' };
    }
    
    // Mark player as escaped
    player.hasEscaped = true;
    this.playersEscaped++;
    
    return { 
      escaped: true, 
      message: 'Successfully escaped!',
      playersEscaped: this.playersEscaped
    };
  }
  
  /**
   * Checks for game over conditions based on the current game mode.
   * @returns {Object} Game over result
   */
  checkGameOverConditions() {
    if (this.isGameOver) {
      return { isGameOver: true, gameMode: this.gameMode };
    }
    
    const totalPlayers = this.players.size;
    const timeUp = this.gameTimer >= this.gameTimeLimit;
    
    let isGameOver = false;
    let message = '';
    let isVictory = false;
    
    // Game over conditions based on game mode
    switch (this.gameMode) {
      case 'coop':
        // All players must escape, or time runs out
        if (this.playersEscaped === totalPlayers || timeUp) {
          isGameOver = true;
          isVictory = this.playersEscaped === totalPlayers;
          message = isVictory ? 
            'All players escaped! Victory!' : 
            'Not everyone escaped. Game over!';
        }
        break;
        
      case 'betrayal':
        // Game ends when one player escapes or time runs out
        if (this.playersEscaped > 0 || timeUp) {
          isGameOver = true;
          isVictory = this.playersEscaped > 0;
          message = isVictory ? 
            'A player has escaped!' : 
            'Time ran out. Game over!';
        }
        break;
        
      case 'pvp':
        // Game ends when all players have either escaped or been trapped, or time runs out
        if (this.playersEscaped + this.playersTrapped === totalPlayers || timeUp) {
          isGameOver = true;
          isVictory = this.playersEscaped > 0;
          message = isVictory ? 
            'Some players escaped!' : 
            'No one escaped. Game over!';
        }
        break;
    }
    
    if (isGameOver) {
      this.isGameOver = true;
    }
    
    return {
      isGameOver,
      gameMode: this.gameMode,
      message,
      isVictory,
      playersEscaped: this.playersEscaped,
      playersTrapped: this.playersTrapped,
      totalPlayers,
      gameTimer: this.gameTimer
    };
  }
  
  /**
   * Applies damage to a player.
   * @param {string} playerId - The player's socket ID
   * @param {number} damage - The amount of damage to apply
   * @returns {Object} The result of the damage application
   */
  applyDamageToPlayer(playerId, damage) {
    const player = this.players.get(playerId);
    if (!player) {
      return { success: false, message: 'Player not found' };
    }
    
    // Apply damage
    player.health -= damage;
    
    // Check if player died
    if (player.health <= 0) {
      player.health = 0;
      player.isTrapped = true;
      this.playersTrapped++;
      
      return { 
        success: true, 
        died: true, 
        health: 0,
        message: 'Player died'
      };
    }
    
    return { 
      success: true, 
      died: false, 
      health: player.health,
      message: 'Player took damage'
    };
  }
  
  /**
   * Sets the maze data.
   * @param {Object} maze - The maze data
   */
  setMaze(maze) {
    this.maze = maze;
  }
  
  /**
   * Gets the current maze data.
   * @returns {Object} The maze data
   */
  getMaze() {
    return this.maze;
  }
  
  /**
   * Gets all players.
   * @returns {Array} Array of player objects
   */
  getPlayers() {
    return Array.from(this.players.values());
  }
  
  /**
   * Gets a specific player.
   * @param {string} playerId - The player's socket ID
   * @returns {Object|undefined} The player object if found
   */
  getPlayer(playerId) {
    return this.players.get(playerId);
  }
  
  /**
   * Gets the current game mode.
   * @returns {string} The game mode
   */
  getGameMode() {
    return this.gameMode;
  }
  
  /**
   * Sets the game mode.
   * @param {string} mode - The game mode to set
   */
  setGameMode(mode) {
    if (['coop', 'betrayal', 'pvp'].includes(mode)) {
      this.gameMode = mode;
    }
  }
  
  /**
   * Gets the current game timer.
   * @returns {number} The game timer in seconds
   */
  getGameTimer() {
    return this.gameTimer;
  }
  
  /**
   * Gets the number of players who have escaped.
   * @returns {number} The number of escaped players
   */
  getPlayersEscaped() {
    return this.playersEscaped;
  }
  
  /**
   * Resets the game state.
   */
  resetGame() {
    this.gameTimer = 0;
    this.playersEscaped = 0;
    this.playersTrapped = 0;
    this.isGameOver = false;
    
    // Reset player states but keep them in the game
    this.players.forEach(player => {
      player.position = this.getRandomSpawnPosition();
      player.health = 100;
      player.hasEscaped = false;
      player.isTrapped = false;
      player.inventory = [];
    });
  }
}

module.exports = GameManager; 

File: Assets/Scripts/Core/GameManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Fusion;
using LabyrinthSurvival.Networking;
using LabyrinthSurvival.Player;
using LabyrinthSurvival.Procedural;
using LabyrinthSurvival.UI;
using LabyrinthSurvival.AI;

namespace LabyrinthSurvival.Core
{
    /// <summary>
    /// Manages the overall game flow, rules, and game modes.
    /// </summary>
    public class GameManager : NetworkBehaviour
    {
        [Header("Game Settings")]
        [SerializeField] private float gameTimeLimit = 600f; // 10 minutes
        [SerializeField] private int maxPlayers = 8;
        [SerializeField] private int monstersPerPlayer = 2;
        [SerializeField] private float monsterSpawnInterval = 60f; // 1 minute
        
        [Header("Game Modes")]
        [SerializeField] private GameMode currentGameMode = GameMode.Coop;
        
        [Header("Prefabs")]
        [SerializeField] private NetworkPrefabRef monsterPrefab;
        [SerializeField] private NetworkPrefabRef itemPrefab;
        
        // Network state
        [Networked] private float GameTimer { get; set; }
        [Networked] private ushort PlayersEscaped { get; set; }
        [Networked] private ushort PlayersTrapped { get; set; }
        [Networked] private NetworkBool IsGameOver { get; set; }
        
        // References
        private MazeGenerator _mazeGenerator;
        private UIManager _uiManager;
        private NetworkManager _networkManager;
        
        // Local state
        private float _nextMonsterSpawnTime;
        private List<NetworkObject> _spawnedMonsters = new List<NetworkObject>();
        
        // Game modes
        public enum GameMode
        {
            Coop,       // All players work together to escape
            Betrayal,   // Only one player can escape
            PvP         // Players can sabotage each other
        }
        
        public override void Spawned()
        {
            base.Spawned();
            
            // Get references
            _mazeGenerator = FindObjectOfType<MazeGenerator>();
            _uiManager = UIManager.Instance;
            _networkManager = NetworkManager.Instance;
            
            // Initialize game state
            GameTimer = 0f;
            PlayersEscaped = 0;
            PlayersTrapped = 0;
            IsGameOver = false;
            
            // Set up monster spawning
            _nextMonsterSpawnTime = monsterSpawnInterval;
            
            // Set up event listeners
            NetworkManager.OnPlayerJoined += OnPlayerJoined;
            NetworkManager.OnPlayerLeft += OnPlayerLeft;
            
            // Start the game
            if (Object.HasStateAuthority)
            {
                StartGame();
            }
        }
        
        public override void FixedUpdateNetwork()
        {
            // Only run game logic on the server
            if (!Object.HasStateAuthority)
                return;
            
            // Update game timer
            GameTimer += Runner.DeltaTime;
            
            // Check for game over conditions
            CheckGameOverConditions();
            
            // Spawn monsters periodically
            if (GameTimer >= _nextMonsterSpawnTime)
            {
                SpawnMonsters();
                _nextMonsterSpawnTime = GameTimer + monsterSpawnInterval;
            }
        }
        
        public override void Render()
        {
            // Update UI
            if (_uiManager != null)
            {
                // Update timer display
                int minutes = Mathf.FloorToInt(GameTimer / 60f);
                int seconds = Mathf.FloorToInt(GameTimer % 60f);
                string timeText = string.Format("{0:00}:{1:00}", minutes, seconds);
                
                // Update objective text based on game mode
                string objectiveText = GetObjectiveText();
                
                // Update UI
                _uiManager.UpdateObjectiveText($"{objectiveText}\nTime: {timeText}");
            }
        }
        
        private void OnDestroy()
        {
            // Clean up event listeners
            if (NetworkManager.Instance != null)
            {
                NetworkManager.OnPlayerJoined -= OnPlayerJoined;
                NetworkManager.OnPlayerLeft -= OnPlayerLeft;
            }
        }
        
        #region Game Flow
        
        /// <summary>
        /// Starts the game.
        /// </summary>
        private void StartGame()
        {
            Debug.Log("Starting game...");
            
            // Generate the maze
            if (_mazeGenerator != null)
            {
                _mazeGenerator.RegenerateMaze();
            }
            
            // Spawn initial monsters
            SpawnMonsters();
            
            // Spawn items
            SpawnItems();
            
            // Show game HUD
            if (_uiManager != null)
            {
                _uiManager.ShowGameHUD();
            }
        }
        
        /// <summary>
        /// Ends the game with the specified outcome.
        /// </summary>
        private void EndGame(bool allPlayersEscaped)
        {
            Debug.Log("Game over!");
            
            // Set game over state
            IsGameOver = true;
            
            // Show game over screen
            if (_uiManager != null)
            {
                string message;
                bool isVictory;
                
                switch (currentGameMode)
                {
                    case GameMode.Coop:
                        isVictory = allPlayersEscaped;
                        message = isVictory ? "All players escaped! Victory!" : "Not everyone escaped. Game over!";
                        break;
                    case GameMode.Betrayal:
                        isVictory = PlayerController.Local != null && PlayerController.Local.HasEscaped;
                        message = isVictory ? "You escaped! Victory!" : "You were betrayed and left behind!";
                        break;
                    case GameMode.PvP:
                        isVictory = PlayerController.Local != null && PlayerController.Local.HasEscaped;
                        message = isVictory ? "You escaped! Victory!" : "You failed to escape!";
                        break;
                    default:
                        isVictory = false;
                        message = "Game over!";
                        break;
                }
                
                _uiManager.ShowGameOver(message, isVictory);
            }
        }
        
        /// <summary>
        /// Checks for game over conditions based on the current game mode.
        /// </summary>
        private void CheckGameOverConditions()
        {
            if (IsGameOver)
                return;
            
            int totalPlayers = Runner.ActivePlayers.Count;
            bool timeUp = GameTimer >= gameTimeLimit;
            
            // Game over conditions based on game mode
            switch (currentGameMode)
            {
                case GameMode.Coop:
                    // All players must escape, or time runs out
                    if (PlayersEscaped == totalPlayers || timeUp)
                    {
                        EndGame(PlayersEscaped == totalPlayers);
                    }
                    break;
                    
                case GameMode.Betrayal:
                    // Game ends when one player escapes or time runs out
                    if (PlayersEscaped > 0 || timeUp)
                    {
                        EndGame(PlayersEscaped > 0);
                    }
                    break;
                    
                case GameMode.PvP:
                    // Game ends when all players have either escaped or been trapped, or time runs out
                    if (PlayersEscaped + PlayersTrapped == totalPlayers || timeUp)
                    {
                        EndGame(PlayersEscaped > 0);
                    }
                    break;
            }
        }
        
        #endregion
        
        #region Spawning
        
        /// <summary>
        /// Spawns monsters based on the number of players.
        /// </summary>
        private void SpawnMonsters()
        {
            int totalPlayers = Runner.ActivePlayers.Count;
            int monstersToSpawn = totalPlayers * monstersPerPlayer - _spawnedMonsters.Count;
            
            for (int i = 0; i < monstersToSpawn; i++)
            {
                // Find a random spawn position away from players
                Vector3 spawnPosition = FindMonsterSpawnPosition();
                
                // Spawn the monster
                NetworkObject monsterObject = Runner.Spawn(
                    monsterPrefab,
                    spawnPosition,
                    Quaternion.identity,
                    null,
                    (runner, obj) => {
                        // Initialize monster
                        MonsterAI monsterAI = obj.GetComponent<MonsterAI>();
                        if (monsterAI != null)
                        {
                            // Any additional setup can go here
                        }
                    }
                );
                
                // Add to list of spawned monsters
                _spawnedMonsters.Add(monsterObject);
            }
        }
        
        /// <summary>
        /// Finds a suitable spawn position for a monster, away from players.
        /// </summary>
        private Vector3 FindMonsterSpawnPosition()
        {
            // Get maze dimensions
            float mazeWidth = 40f; // Assuming 20x20 maze with 2 unit cells
            float mazeHeight = 40f;
            
            // Try to find a position away from players
            for (int attempts = 0; attempts < 10; attempts++)
            {
                // Random position within maze bounds
                Vector3 position = new Vector3(
                    Random.Range(2f, mazeWidth - 2f),
                    0.5f,
                    Random.Range(2f, mazeHeight - 2f)
                );
                
                // Check distance to all players
                bool isTooClose = false;
                foreach (var player in Runner.ActivePlayers)
                {
                    PlayerController playerController = Runner.GetPlayerObject(player)?.GetComponent<PlayerController>();
                    if (playerController != null)
                    {
                        float distance = Vector3.Distance(position, playerController.transform.position);
                        if (distance < 10f) // Minimum distance from players
                        {
                            isTooClose = true;
                            break;
                        }
                    }
                }
                
                if (!isTooClose)
                {
                    return position;
                }
            }
            
            // Fallback: return a random position
            return new Vector3(
                Random.Range(2f, mazeWidth - 2f),
                0.5f,
                Random.Range(2f, mazeHeight - 2f)
            );
        }
        
        /// <summary>
        /// Spawns items throughout the maze.
        /// </summary>
        private void SpawnItems()
        {
            // Get maze dimensions
            float mazeWidth = 40f; // Assuming 20x20 maze with 2 unit cells
            float mazeHeight = 40f;
            
            // Spawn a number of items
            int itemsToSpawn = 20; // Adjust as needed
            
            for (int i = 0; i < itemsToSpawn; i++)
            {
                // Random position within maze bounds
                Vector3 position = new Vector3(
                    Random.Range(2f, mazeWidth - 2f),
                    0.5f,
                    Random.Range(2f, mazeHeight - 2f)
                );
                
                // Random item type
                int itemType = Random.Range(0, 4); // 0-3 for different item types
                
                // Spawn the item
                Runner.Spawn(
                    itemPrefab,
                    position,
                    Quaternion.identity,
                    null,
                    (runner, obj) => {
                        // Initialize item
                        ItemPickup itemPickup = obj.GetComponent<ItemPickup>();
                        if (itemPickup != null)
                        {
                            itemPickup.SetItemType((ItemPickup.ItemType)itemType);
                        }
                    }
                );
            }
        }
        
        #endregion
        
        #region Event Handlers
        
        /// <summary>
        /// Called when a player joins the game.
        /// </summary>
        private void OnPlayerJoined(PlayerRef player, NetworkObject playerObject)
        {
            Debug.Log($"Player {player} joined the game");
            
            // Additional setup for the player can go here
        }
        
        /// <summary>
        /// Called when a player leaves the game.
        /// </summary>
        private void OnPlayerLeft(PlayerRef player)
        {
            Debug.Log($"Player {player} left the game");
            
            // Handle player leaving based on game mode
            if (Object.HasStateAuthority)
            {
                // In co-op mode, if a player leaves, they're considered trapped
                if (currentGameMode == GameMode.Coop)
                {
                    PlayersTrapped++;
                    CheckGameOverConditions();
                }
            }
        }
        
        #endregion
        
        #region Helper Methods
        
        /// <summary>
        /// Gets the objective text based on the current game mode.
        /// </summary>
        private string GetObjectiveText()
        {
            switch (currentGameMode)
            {
                case GameMode.Coop:
                    return "Objective: Work together to find the exit and escape!";
                case GameMode.Betrayal:
                    return "Objective: Be the first to find the exit and escape!";
                case GameMode.PvP:
                    return "Objective: Find the exit and escape, or prevent others from escaping!";
                default:
                    return "Objective: Find the exit and escape the labyrinth!";
            }
        }
        
        /// <summary>
        /// Called when a player escapes the maze.
        /// </summary>
        public void PlayerEscaped(PlayerRef player)
        {
            if (Object.HasStateAuthority)
            {
                PlayersEscaped++;
                Debug.Log($"Player {player} escaped! ({PlayersEscaped}/{Runner.ActivePlayers.Count})");
                
                // Check game over conditions
                CheckGameOverConditions();
            }
        }
        
        /// <summary>
        /// Called when a player is trapped (e.g., killed by a monster).
        /// </summary>
        public void PlayerTrapped(PlayerRef player)
        {
            if (Object.HasStateAuthority)
            {
                PlayersTrapped++;
                Debug.Log($"Player {player} trapped! ({PlayersTrapped}/{Runner.ActivePlayers.Count})");
                
                // Check game over conditions
                CheckGameOverConditions();
            }
        }
        
        #endregion
    }
    
    /// <summary>
    /// Represents an item that can be picked up by players.
    /// </summary>
    public class ItemPickup : NetworkBehaviour, IInteractable
    {
        [SerializeField] private MeshRenderer meshRenderer;
        [SerializeField] private Material[] itemMaterials;
        
        [Networked] private ItemType _itemType { get; set; }
        
        public enum ItemType
        {
            Key,
            Weapon,
            Health,
            Tool
        }
        
        public override void Spawned()
        {
            base.Spawned();
            
            // Set material based on item type
            if (meshRenderer != null && itemMaterials != null && itemMaterials.Length > (int)_itemType)
            {
                meshRenderer.material = itemMaterials[(int)_itemType];
            }
        }
        
        public void SetItemType(ItemType type)
        {
            _itemType = type;
        }
        
        public void Interact(PlayerController player)
        {
            Debug.Log($"Player picked up {_itemType}");
            
            // Add item to player's inventory
            // player.AddItem(_itemType);
            
            // Despawn the item
            if (Object.HasStateAuthority)
            {
                Runner.Despawn(Object);
            }
        }
    }
} 